src/syslog_udp_sender.erl

```diff
@@ -0,0 +1,129 @@
+%%%-------------------------------------------------------------------
+%%% @author tihon
+%%% @copyright (C) 2016, <COMPANY>
+%%% @doc
+%%%
+%%% @end
+%%% Created : 17. Jun 2016 15:18
+%%%-------------------------------------------------------------------
+-module(syslog_udp_sender).
+-author("tihon").
+
+-behaviour(gen_server).
+
+-include("syslog.hrl").
+
+%% API
+-export([start_link/0]).
+
+%% gen_server callbacks
+-export([init/1,
+  handle_call/3,
+  handle_cast/2,
+  handle_info/2,
+  terminate/2,
+  code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+-record(state,
+{
+  hibernate_timer :: timer:tref() | undefined,
+  socket :: inet:socket(),
+  dest_host :: inet:ip_address() | inet:hostname(),
+  dest_port :: inet:port_number()
+}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the server
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(start_link() ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+  gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @private
+%%--------------------------------------------------------------------
+init({Host, Port}) ->
+  {ok, Socket} = gen_udp:open(0, [binary]),
+  {ok, #state{dest_host = Host, dest_port = Port, socket = Socket}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%%--------------------------------------------------------------------
+handle_call(_Request, _From, State) ->
+  {reply, ok, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%%--------------------------------------------------------------------
+handle_cast(_Request, State) ->
+  {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%%--------------------------------------------------------------------
+handle_info({send, Report, Protocol}, State = #state{socket = S, dest_host = H, dest_port = P}) ->
+  [send_datagram(S, H, P, Protocol:to_iolist(R)) || R <- split(Report)],
+  syslog_udp_sup:return_worker(),
+  {noreply, need_timer(Report, State)};
+handle_info({udp_closed, S}, #state{socket = S}) ->
+  {error, {udp_closed, S}};
```

Comment:
Invalid `gen_server` return. Should probably be something like `{stop, {error, {udp_closed}}, S}`.